docs/spec.rst

```diff
@@ -0,0 +1,501 @@
+
+Semantics
+---------
+
+In this section, we discuss how subtyping should play out.
+
+Subtyping and type conversion
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Our nullability annotations produce the following subtyping hierarchy
+[`#80 <https://github.com/jspecify/jspecify/issues/80>`__]:
+
+| ``Foo`` (written in the context of ``@DefaultNonNull``)
+| ``<: Foo`` (written outside the context of ``@DefaultNonNull``)
+| ``<: @Nullable Foo``
+
+It can be useful to conceptualize these similarly to `3-valued
+logic <https://en.wikipedia.org/wiki/Three-valued_logic>`__
+[`#33 <https://github.com/jspecify/jspecify/issues/33>`__]. While that’s
+workable, it can be useful to instead define “unspecified nullness” as
+an existential quantification over the other two (which logically still
+validates the above hierarchy). Two unrelated occurrences of unspecified
+nullness can sometimes be represented as two different variables,
+similar to how the Java type system handles wildcards.
+
+The above rules make ``@Nullable Object`` the top (least precise) type.
+(Note that ``null`` is *not* the bottom type.) Here are some more
+examples of subtyping, with types written in the context of
+``@DefaultNonNull``:
+
+-  ``String <: Object <: @Nullable Object``
+-  ``String <: @Nullable String <: @Nullable Object``
+-  ``null <: @Nullable String <: @Nullable Object``
+
+For a given set of nullability specifiers, we can define their **glb**
+(*greatest lower bound*) as a specifier from the given set that is at
+least as specific as all others. Similarly, the **lub** (*least upper
+bound*) of a set of specifiers is a specifier from the set that is at
+most as specific as all others.
+
+Finally, a **type check** (e.g., to determine assignability) for a pair
+of `augmented
+types <https://docs.google.com/document/d/1KQrBxwaVIPIac_6SCf--w-vZBeHkTvtaqPSU_icIccc/edit#bookmark=id.367l48xhsikk>`__
+includes validating **both** (a) a type check of the `base
+types <https://docs.google.com/document/d/1KQrBxwaVIPIac_6SCf--w-vZBeHkTvtaqPSU_icIccc/edit#bookmark=kix.e98jyoqkxvjd>`__,
+handled by the off-the-shelf compiler, and (b) a type check for
+nullability.
+
+-  This specification additionally encourages allowing **unsound
+   nullability subtyping rules** (optionally with warnings, similar to
+   “unchecked conversions” in Java generics) involving unspecified
+   nullness, such as the following
+   [`#33 <https://github.com/jspecify/jspecify/issues/33>`__]:
+
+   -  nullable as a subtype of unspecified nullness
+   -  unspecified nullness as a subtype of non-null
+   -  unspecified nullness as a subtype of unspecified nullness
+
+-  For usability reasons, tools that want to generate warnings/errors
+   when applying unsound rules like those above are **strongly**
+   encouraged to permit the suppression of these warnings without
+   suppressing other soundness violations. (In fact, we expect this to
+   be the desirable default behavior.)
```

Comment:
this is probably ambiguously worded in the proposal.  I think what's meant here is that we expect most tools to not even warn about these unsound conversions.

Reply:
I rewrote this to try to clarify the various options. I don't love it, but maybe it's at least clearer about what the options are?